### ISSUE TITLE
Questions on how the code works.

### ISSUE THREAD

**Issue by mhwang003** [facebookresearch/silk#41]
Hi, I have 2 questions regarding the code.

The first question is how the flow module works. I noticed that in silk.py You use [this line](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/models/silk.py#L203) to define the loss function but I was not able to understand how it is connected to  [the loss function] (https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/losses/info_nce/loss.py#L236). 
Also It seemed to me that you can add functions by using the [define_transition](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/models/silk.py#L137) method but I do not understand how this works either.

The second question is about how you save the losses from training to tfevents. It seems the losses are saved [here](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/models/silk.py#L333) but I was not able to find where the logging function was defined.

Any pointers to how the code works would be appreciated! Thanks in advance.

**Comment by mhwang003** [comment]
Another question I have is, is it possible to turn off JIT while using JAX? I have been having trouble debugging my code I have implemented in the loss function. Thanks.

**Comment by gleize** [comment]
Hi @mhwang003,

> The first question is how the flow module works.

The `Flow` module implements a simple python-based execution DAG.
* Each node represents data or state.
* Each edge represents a function to run to obtain the data.

For example, the following 2-nodes, 1-edge DAG indicates how to obtain the data `feature` from the data `image`, using the function `model_fn`.
```
          model_fn
image  ------------->  feature
```

In our flow library, this would be defined by a call like this :
```python
        self.flow.define_transition(
            "feature",                             # output
            model_fn,                              # function
            "image",                               # function inputs, as node name
        )
```
Note that the function `model_fn` will not be executed upon calling `define_transition`. Here, we just tell the flow library that if it wants to obtain `feature` from `image`, it can by running the function `model_fn` on whatever data is stored in `image` during runtime.

Now, you can imagine that you can have more complex flow with multiple connections. An execution DAG is really handy when someone wants to code a complex pipeline, but doesn't know what intermediate state is going to be requested at runtime (which is the case of SiLK). So, instead of coding multiple versions of the same code that stops at different states in the graph, an execution DAG allows you to define the DAG only once, and ask the DAG which output you request on demand.

For example, imagine you have this execution DAG :
```
   fn_0      fn_1
a ------> b ------> c
          | ------> d
             fn_2
```

During runtime, I can ask any of those states. But also, the runtime will be smart enough to only run the strict necessary transitions to obtain the result. For example, if I request the data `d`, the runtime exection will essentially run :
```
b = fn_0(a)
d = fn_2(b)
```
and completely skip the computation of `c` (which makes the runtime efficient).

You can look a bit more at our unit tests [here](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/flow_test.py) to learn more about `Flow`.

> You use [this line](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/models/silk.py#L203) to define the loss function but I was not able to understand how it is connected to [the loss function]

This [call](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/models/silk.py#L203) essentially creates a function, that, when you call it, will return the states `contextual_descriptor_loss`, `acontextual_descriptor_loss`, `keypoint_loss`, `precision` and `recall` from the defined flow.

You can notice the loss transition is defined [here](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/lib/models/silk.py#L175C24-L175C24), which calls the loss function configured [here](https://github.com/facebookresearch/silk/blob/0a7da0168ba4fe2f4e49ebff0980ebcff0d2a1b3/etc/models/silk-default.yaml#L1).

> The second question is about how you save the losses from training to tfevents.

The `log()` method is implemented by [PyTorch Lightning](https://lightning.ai/) and available to our class since we subclass `pl.LightningModule`.

> Another question I have is, is it possible to turn off JIT while using JAX? 

[Yes](https://jax.readthedocs.io/en/latest/_autosummary/jax.disable_jit.html).

**Comment by mhwang003** [comment]
Thank you!  I have a better understanding of the code now. Much thanks.

**Comment by mhwang003** [comment]
Hi, I tried using `with jax.disable_jit(): `in the loss funtion in loss.py but I'm still getting a `jax._src.errors.TracerArrayConversionError `when trying to convert the jax.numpy array into a regular numpy array for debugging and plotting. I do not understand jax very well, is this error due to JIT compilation of a function or another reason? And is it possible to disable JIT compilation in the whole code?

**Comment by gleize** [comment]
Hi @mhwang003,

Could you share where you added the `with jax.disable_jit():` and where you did the numpy array conversion ?
I'll try to reproduce the issue and see what's going on.

**Comment by gleize** [comment]
No activity. Closing now.
